### ISSUE TITLE
SFT and RL training codes and scripts for olmo3?

### ISSUE THREAD

**Issue by rasoolfa** [allenai/OLMo-core#488]
hello,

Thanks for releasing such a detailed paper and work. it is a great asset indeed. 

I have a couple of questions about the post-training though:

1. In A.4.1 of the paper, you mention that you switched from open-instruct to olmo-core for post-training in olmo3. However, I can’t find any scripts for SFT, DPO, or RL training. Could you point me to those?

2. It seems that pre-training was done using FSDP2, and that you switched away from DeepSpeed(?). My understanding is that open-instruct is solely based on DeepSpeed. Could you clarify which framework you used for post-training (SFT, DPO, and RL) in olmo-3?

thanks,
rasool

**Comment by epwalsh** [comment]
Hi, we used olmo-core for SFT but not DPO or RL. You can find those scripts here: https://github.com/allenai/OLMo-core/tree/main/src/scripts/official/OLMo3 (the ones named `*-midtrain*`). Yes, pretraining used FSDP2 and so did SFT. All post-training done directly in open-instruct (so DPO and RL) used DeepSpeed.

**Comment by rasoolfa** [comment]
thanks for your response. 

Just to confirm my understanding based on your reply, *-midtrain* scrips are SFT script, correct? 
Also, to follow up: is the main source of 8x faster  SFT training simply switching from DeepSpeed to FSDP2, or are there more to it?

Regarding open-instruct, is this correct path for your RL and DPO's training scripts?
https://github.com/allenai/open-instruct/tree/main/scripts/train/olmo3

**Comment by epwalsh** [comment]
> Just to confirm my understanding based on your reply, `-midtrain` scripts are SFT script, correct?

Oops no sorry, those scripts are for the various pretraining stages only, although they could be adapted for SFT by changing the data sources. The actual post-training SFT scripts can be found here: https://github.com/allenai/OLMo-core/tree/jacobm-final-merge-sft/src/scripts/train/sft

But note that only Ai2 employees/collaborators can run those scripts as-is at the moment since they assume access to internal compute and data storage. We're working on creating versions of those scripts for the general public, which we'll publish on `main`.

> Also, to follow up: is the main source of 8x faster SFT training simply switching from DeepSpeed to FSDP2, or are there more to it?

There's more to it. To name a few things: `torch.compile()`, flash-attn 2, and async checkpointing and metric collection. We talked more about olmo-core's features in the OLMo 2 paper.

> Regarding open-instruct, is this correct path for your RL and DPO's training scripts?

Yes.

**Comment by rasoolfa** [comment]
thanks again for your reply.
